Userland/Libraries/LibJS/Runtime/AbstractOperations.cpp

```diff
@@ -0,0 +1,164 @@
+    if (vm.exception())
+        return MarkedValueList { heap };
+    auto list = MarkedValueList { heap };
+    for (size_t i = 0; i < length; ++i) {
+        auto index_name = String::number(i);
+        auto next = array_like.get(index_name).value_or(js_undefined());
+        if (vm.exception())
+            return MarkedValueList { heap };
+        if (check_value) {
+            auto result = check_value(next);
+            if (result.is_error()) {
+                vm.throw_exception<TypeError>(global_object, result.release_error());
+                return MarkedValueList { heap };
+            }
+        }
+        list.append(next);
+    }
+    return list;
+}
+
+// 7.3.22 SpeciesConstructor ( O, defaultConstructor ), https://tc39.es/ecma262/#sec-speciesconstructor
+Function* species_constructor(GlobalObject& global_object, Object const& object, Function& default_constructor)
+{
+    auto& vm = global_object.vm();
+    auto constructor = object.get(vm.names.constructor).value_or(js_undefined());
+    if (vm.exception())
+        return nullptr;
+    if (constructor.is_undefined())
+        return &default_constructor;
+    if (!constructor.is_object()) {
+        vm.throw_exception<TypeError>(global_object, ErrorType::NotAConstructor, constructor.to_string_without_side_effects());
+        return nullptr;
+    }
+    auto species = constructor.as_object().get(vm.well_known_symbol_species()).value_or(js_undefined());
+    if (species.is_nullish())
+        return &default_constructor;
+    if (species.is_constructor())
+        return &species.as_function();
+    vm.throw_exception<TypeError>(global_object, ErrorType::NotAConstructor, species.to_string_without_side_effects());
+    return nullptr;
+}
+
+// 7.3.24 GetFunctionRealm ( obj ), https://tc39.es/ecma262/#sec-getfunctionrealm
+GlobalObject* get_function_realm(GlobalObject& global_object, Function const& function)
+{
+    auto& vm = global_object.vm();
+
+    // FIXME: not sure how to do this currently.
+    // 2. If obj has a [[Realm]] internal slot, then
+    //     a. Return obj.[[Realm]].
+    if (is<BoundFunction>(function)) {
+        auto& bound_function = static_cast<BoundFunction const&>(function);
+        auto& target = bound_function.target_function();
+        return get_function_realm(global_object, target);
+    }
+    if (is<ProxyObject>(function)) {
+        auto& proxy = static_cast<ProxyObject const&>(function);
+        if (proxy.is_revoked()) {
+            vm.throw_exception<TypeError>(global_object, ErrorType::ProxyRevoked);
+            return nullptr;
+        }
+        auto& proxy_target = proxy.target();
+        // NOTE: I have no idea how this is _supposed_ to work - the spec has this as first step:
+        // 1. Assert: ! IsCallable(obj) is true.
+        // and yet they recursively pass obj.[[ProxyTarget]] to GetFunctionRealm(),
+        // even though it could be any object?! Either way, this should do the trick.
```

Comment:
A proxy whose target is a function will have a `[[Call]]` internal slot, and will therefore return true from `IsCallable` (see [ProxyCreate](https://tc39.es/ecma262/#sec-proxycreate)).

Reply:
Great! So `GetFunctionRealm` is guaranteed to only ever be called on a Proxy with a callable target, that makes sense. Thanks :)